libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt

```diff
@@ -1229,11 +1272,23 @@ class MapboxNavigation @VisibleForTesting internal constructor(
     }
 
     private fun reroute() {
-        rerouteController?.reroute { routes ->
-            directionsSession.setRoutes(
-                routes,
-                routesUpdateReason = RoutesExtra.ROUTES_UPDATE_REASON_REROUTE
+        val controller = rerouteController
+        if (controller is NavigationRerouteController) {
+            controller.reroute(
+                NavigationRerouteController.RoutesCallback { routes ->
+                    directionsSession.setRoutes(
+                        routes,
+                        routesUpdateReason = RoutesExtra.ROUTES_UPDATE_REASON_REROUTE
+                    )
+                }
             )
+        } else {
+            rerouteController?.reroute { routes ->
+                directionsSession.setRoutes(
+                    routes.toNavigationRoutes(),
+                    routesUpdateReason = RoutesExtra.ROUTES_UPDATE_REASON_REROUTE
+                )
+            }
```

Comment:
`if` could be replaced by polymorphism if you wrap the old reroute controller in an adapter so that `MapboxNavigation` always keeps reference to `NavigationRerouteController`

```koltin
class OldRerouteControllerAdapter(
    private val oldRerouteController: RerouteController
) : NavigationRerouteController, RerouteController by oldRerouteController {

    override fun reroute(callback: NavigationRerouteController.RoutesCallback) {
        oldRerouteController.reroute(object : RerouteController.RoutesCallback {
            override fun onNewRoutes(routes: List<DirectionsRoute>) {
                callback.onNewRoutes(routes.toNavigationRoutes())
            }
        })
    }
}
```